solidity-bindgen-macros/src/abi_gen.rs

```diff
@@ -111,72 +114,96 @@ pub fn to_rust_name(type_name: &str, eth_name: &str, i: usize) -> String {
-            let eth_name = &function.name;
-            let rust_name = Ident::new(&to_rust_name("function", eth_name, 0), span);
-
-            // Get the types and names of parameters
-            let params_in = function.inputs.iter().enumerate().map(|(i, param)| {
-                let name = Ident::new(&to_rust_name("input", &param.name, i), span);
-                let t = param_type(&param.r#type);
-                quote! {
-                    #name: #t
-                }
-            });
-
-            let params = function.inputs.iter().enumerate().map(|(i, param)| {
-                let name = Ident::new(&to_rust_name("input", &param.name, i), span);
-                quote! { #name }
-            });
-            let params = if function.inputs.len() == 1 {
-                quote! { #(#params)* }
-            } else {
-                quote! { (#(#params),*) }
-            };
-
-            let transaction = matches!(
-                function.state_mutability,
-                StateMutability::Nonpayable | StateMutability::Payable
-            );
-            let method = if transaction { "send" } else { "call" };
-            let method = Ident::new(method, Span::call_site());
-
-            let ok = if transaction {
-                // Despite information in the ABIs to the contrary, there aren't
-                // really outputs for web3 send fns. The outputs that are
-                // available aren't returned by these APIs, but are only made
-                // available to contracts calling each other. 🤷
-                //
-                // All you can get is a receipt. So, the way to get something
-                // like a return value would be to check for events emitted or
-                // to make further queries for data.
-                quote! { ::web3::types::TransactionReceipt }
-            } else {
-                match function.outputs.len() {
-                    0 => quote! { ::solidity_bindgen::internal::Empty },
-                    1 => {
-                        let t = param_type(&function.outputs[0].r#type);
-                        quote! { #t }
-                    }
-                    _ => {
-                        let types = function.outputs.iter().map(|o| {
-                            let t = param_type(&o.r#type);
-                            quote! { #t }
-                        });
-
-                        quote! { (#(#types),*) }
+pub fn fn_from_abi(function: &Function) -> TokenStream {
+    let eth_name = &function.name;
+    let rust_name = ident(to_rust_name("function", eth_name, 0));
+
+    // Get the types and names of parameters
+    let params_nesting = if function.inputs.len() > 1 { 1 } else { 0 };
+    let params_in = function.inputs.iter().enumerate().map(|(i, param)| {
+        let name = ident(to_rust_name("input", &param.name, i));
+        let (t, nesting) = param_type(&param.kind);
+
+        // We have to have a branch here because Tokenize isn't implemented for
```

Comment:
The background of the incompatibility and next steps for impl tokenize for nested tuples aren't 100% clear to me from this comment.